### DISCUSSION TITLE
Streaming API

### DISCUSSION THREAD

**Discussion by marsianer** [ical4j/ical4j#762] [Q&A]
Is it possible to do event-based processing so that the entire calendar does not have to be loaded into memory?
Unfortunately, my calendar contains many attachments (GPX recordings of workouts) that make the calendar grow.
On my Android smartphone, this unfortunately leads to an OOM termination.

[DAVx⁵ / OOM termination](https://github.com/bitfireAT/davx5-ose/issues/1161)

This method loads the entire calendar into memory.
```
FileInputStream fin = new FileInputStream("calendar.ics"); 
CalendarBuilder builder = new CalendarBuilder();
Calendar calendar = builder.build(fin);
```

Is there a method like this?
```
File file = new File("calendar.ics");
CalendarReader calendarReader = new CalendarReader(file);
Calendar calendar;
while ((calendar = calendarReader.readNext()) != null){
  // ...
}
calendarReader.close();
```

**Comment by benfortuna** [comment]
Yes, if you want to implement streaming, or even conditional parsing (e.g. skip large attachments), you can implement your own [ContentHandler](https://github.com/ical4j/ical4j/blob/develop/src/main/java/net/fortuna/ical4j/data/ContentHandler.java).

This is provided to the [CalendarParser](https://github.com/ical4j/ical4j/blob/develop/src/main/java/net/fortuna/ical4j/data/CalendarParser.java), to process each line. The `CalendarBuilder` just encapsulates a `DefaultContentHandler` and `CalendarParser` for convenience.

```java
File file = new File("calendar.ics");
// unfolding reader handles split lines..
UnfoldingReader calendarReader = new UnfoldingReader(file);

CalendarParser parser = CalendarParserFactory.getInstance().get();
ContentHandler handler = ...

parser.parse(calendarReader, handler);
```

**Comment by marsianer** [comment]
Are there any disadvantages to this, in contrast to loading the entire calendar into memory.
If there are no disadvantages, I would pass this information on to the DavX5 maintainers.

Unfortunately, I haven't been able to use DavX5 for months due to OOM.

**Reply by benfortuna** [reply]
No major disadvantages, other than you can't use the `CalendarBuilder` and `DefaultContentHandler`, so would need to replicate their functionality as needed. So good to have an understanding of these two classes first.

**Reply by rfc2822** [reply]
In case of Android, I think it won't work because processing of properties can depend on other properties, and there's no defined order.

For instance DURATION/DTEND and RRULE: Android requires that for recurring events, an event must have a DURATION, but not a DTEND.

So we have to look at the recurrence rules before we know how to process the DTEND.

If we just use a streaming API, the first property we stumble upon may be DTEND, and then large inline attachments, and then the recurrence rules.

**Reply by benfortuna** [reply]
Ah ok, I was thinking more that you can skip the properties with large attachments using a custom content handler. 

But yeah if you want to use a partially parsed calendar that would be problematic. You need to load all the essential properties before further operations.

**Reply by marsianer** [reply]
Is it possible to parse the components line by line?

The first step is to read all essential properties, followed by the individual components. I think this is still more efficient than loading everything into memory. And in case the essential properties are always located in the beginning of an ICS file, then multiple parsing shouldn't be necessary.

```
for (Component component: calendar.getComponents()){
  // Only one component is kept in memory at a time.
}
```

An OOM would only occur if very large attachments are contained within a component.
But does anyone do that?

**Reply by benfortuna** [reply]
When you parse a calendar file/data stream the whole calendar will be in memory (i.e. all components). Typically this is not an issue as I have not seen examples of large inline attachments, but if you have this concern you coud try the following:

1. Use a custom property factory to convert inline attachments to references (i.e. URL-based) by writing the content to an external source (local files, object storage, etc.). If you process one at at time it should not result in OOM and you can then load the entire calendar in memory.
2. Use a custom content handler to read a single component from a calendar file, as most components are independent of others (although it is possible for multiple components to share the same UID property, in which case they should be processed as a group).

The current implementation of the `ATTACH` property uses a byte array to store inline attachments, however I might explore if this can be enhanced to support memory-mapped files, which would allow processing of large inline attachments without requiring the entire file in memory.

**Reply by benfortuna** [reply]
@marsianer  I have just implemented an experimental use of file-mapped bytebuffers to support larger attachment binaries.

You can try it via this snapshot dependency:

```xml
<dependency>
    <groupId>org.mnode.ical4j</groupId>
    <artifactId>ical4j</artifactId>
    <version>4.2.0-develop-SNAPSHOT</version>
</dependency>
```
